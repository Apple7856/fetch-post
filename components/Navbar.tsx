import { alpha, AppBar, Avatar, Badge, InputBase, makeStyles, Toolbar, Typography } from '@material-ui/core';
import { Cancel, Mail, Notifications, Search } from '@material-ui/icons';
import React, { useState } from 'react';

const useStyle = makeStyles((theme) => ({
  toolbar: {
    display: "flex",
    justifyContent: "space-between",
  },
  logo: {
  },
  search: {
    display: "flex",
    alignItems: "center",
    backgroundColor: alpha(theme.palette.common.white, 0.15),
    '&:hover': {
      backgroundColor: alpha(theme.palette.common.white, 0.25),
    },
    borderRadius: theme.shape.borderRadius,
    width: "30%",
  },
  input: {
    color: "white",
    marginLeft: theme.spacing(2),
  },
}));



const Navbar: React.FC = () => {
  const classes = useStyle();

  function getSearchData(e:any) {
    const inputVal = e.target.value.toLowerCase();
    const tableRow = document.getElementsByClassName('tableRow');
    Array.from(tableRow).forEach(function (element) {
      let title = element.getElementsByClassName("title")[0].innerHTML;
      let url = element.getElementsByClassName('url')[0].innerHTML;
      let author = element.getElementsByClassName('author')[0].innerHTML;
      if (title.includes(inputVal) || url.includes(inputVal) || author.includes(inputVal)) {
        element.style.display = "block";
      }
      else {
        element.style.display = "none";
      }
  })
  }
  return (
    <AppBar position='fixed'>
      <Toolbar className={classes.toolbar}>
        <Typography variant='h6' className={classes.logo}>
          Fetch-Post
        </Typography>
        <div className={classes.search}>
          <Search />
          <InputBase placeholder='Search...' className={classes.input} onChange={(e)=>getSearchData(e)} />
        </div>
      </Toolbar>
    </AppBar>
  )
}

export default Navbar;
