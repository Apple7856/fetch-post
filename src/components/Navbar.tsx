import { alpha, AppBar, Avatar, Badge, InputBase, makeStyles, Toolbar, Typography } from '@material-ui/core';
import { Cancel, Mail, Notifications, Search } from '@material-ui/icons';
import React, { useState } from 'react';

const useStyle = makeStyles((theme) => ({
  toolbar: {
    display: "flex",
    justifyContent: "space-between",
  },
  logo: {
  },
  search: {
    display: "flex",
    alignItems: "center",
    backgroundColor: alpha(theme.palette.common.white, 0.15),
    '&:hover': {
      backgroundColor: alpha(theme.palette.common.white, 0.25),
    },
    borderRadius: theme.shape.borderRadius,
    width: "30%",
  },
  input: {
    color: "white",
    marginLeft: theme.spacing(2),
  },
}));

type SearchProps = {
  handleSearch: (e: any) => void
}

const Navbar = ({handleSearch}:SearchProps) => {
  const classes = useStyle();

  return (
    <AppBar position='fixed'>
      <Toolbar className={classes.toolbar}>
        <Typography variant='h6' className={classes.logo}>
          Fetch-Post
        </Typography>
        <div className={classes.search}>
          <Search />
          {/* <input placeholder='Search...' data-testid="searchBar" value="SearchData" className={classes.input} onChange={(e)=>handleSearch(e)} type="text" /> */}
          <InputBase placeholder='Search...' data-testid="searchBar" value="SearchData" className={classes.input} onChange={(e)=>handleSearch(e)} />
        </div>
      </Toolbar>
    </AppBar>
  )
}

export default Navbar;
