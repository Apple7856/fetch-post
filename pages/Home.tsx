import React, { useEffect, useState } from 'react';
import { Typography, makeStyles, TableContainer, Table, TableHead, TableRow, TableCell, Paper, TableBody, Modal } from '@material-ui/core'
import axios from 'axios';
import Navbar from '../components/Navbar';

const useStyle = makeStyles({
    home: {
        marginTop: "62px",
        height: '89.5vh',
        display: 'flex',
        flexDirection: 'column',
        alignItems: "center",
    },
    tableContainer: {
        maxWidth: "1200px",
        marginTop: "10px"
    },
    table: {

    },
    heading: {
        fontWeight: "bold",
    },
    paper: {
        position: 'absolute',
        width: 600,
        height: 300,
        backgroundColor: "white",
        border: '2px solid #000',
        overflow: "hidden"
    },
    modalBox: {
        display: "flex",
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "center"
    }
});

interface state {
    data: Array<string | number>,
    value: boolean,
    count: number,
    open: boolean,
    rowData: any,
}

const Home: React.FC<state> = () => {

    const [data, setData] = useState([]);
    const [count, setCount] = useState(0);
    const [value, setValue] = useState(true);
    const [open, setOpen] = useState(false);
    const [rowData, setRowData] = useState("");
    const classes = useStyle();

    const onScroll = (e: any) => {
        const bottom = Math.trunc(e.target.scrollHeight - e.target.scrollTop) === e.target.clientHeight;
        if (bottom) {
            setValue(false);
            setCount(count + 1);
        }
    }

    const timer = () => {
        value ? setCount(count + 1) : ""
    };

    useEffect(() => {
        console.log(count);
        axios.get(`https://hn.algolia.com/api/v1/search_by_date?tags=story&page=${count}`)
            .then((res) => {
                setData(data.concat(res.data.hits));
            })
            .catch((error) => {
                console.log(error);
            });
        if (value) {
            const id = setInterval(timer, 10000);
            return () => clearInterval(id);
        }
    }, [count]);

    const handleClose = () => {
        setOpen(false);
    }
    const handleOpen = (x: any) => {
        const findData: any = data.find((item: any) => item.objectID === x);
        setRowData(findData);
        setOpen(true);
    }
    return (
        <Typography component="div" className={classes.home}>
            <Navbar />
            <TableContainer component={Paper} className={classes.tableContainer} onScroll={onScroll}>
                <Table className={classes.table} aria-label="simple table">
                    <TableHead>
                        <TableRow>
                            <TableCell align="center" className={classes.heading}>Sr. No.</TableCell>
                            <TableCell align="center" className={classes.heading}>Title</TableCell>
                            <TableCell align="center" className={classes.heading}>URL</TableCell>
                            <TableCell align="center" className={classes.heading}>Created_at</TableCell>
                            <TableCell align="center" className={classes.heading}>Author</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {data.map((row: any, i) => (
                            <TableRow key={i} className='tableRow' onClick={(x) => handleOpen(row.objectID)}>
                                <TableCell align="center" className="url">{i + 1}</TableCell>
                                <TableCell component="th" scope="row" align="center" className="title">
                                    {row.title}
                                </TableCell>
                                <TableCell align="center" className="url">{row.url}</TableCell>
                                <TableCell align="center">{row.created_at}</TableCell>
                                <TableCell align="center" className="author">{row.author}</TableCell>
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
            <Modal
                open={open}
                onClose={handleClose}
                aria-labelledby="simple-modal-title"
                aria-describedby="simple-modal-description"
                className={classes.modalBox}
            >
                <div className={classes.paper}>
                    <p id="simple-modal-description">
                        {JSON.stringify(rowData)}
                    </p>
                </div>
            </Modal>
        </Typography>
    )
}

export default Home;
