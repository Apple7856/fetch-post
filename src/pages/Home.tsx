import React, { useEffect, useState } from 'react';
import { Typography, makeStyles, TableContainer, Table, TableHead, TableRow, TableCell, Paper, TableBody, Modal } from '@material-ui/core'
import axios from 'axios';
import Navbar from '../components/Navbar';

const useStyle = makeStyles({
    home: {
        marginTop: "62px",
        height: '89.5vh',
        display: 'flex',
        flexDirection: 'column',
        alignItems: "center",
    },
    tableContainer: {
        maxWidth: "1200px",
        marginTop: "10px"
    },
    table: {

    },
    heading: {
        fontWeight: "bold",
    },
    paper: {
        position: 'absolute',
        width: 600,
        height: 300,
        backgroundColor: "white",
        border: '2px solid #000',
        overflow: "hidden"
    },
    modalBox: {
        display: "flex",
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "center"
    }
});


const Home = () => {

    const [data, setData] = useState<Array<string | number>>([]);
    const [searchData, setSearchData] = useState<Array<string | number>>([]);
    const [displayData, setDisplayData] = useState<boolean>(false);
    const [count, setCount] = useState<number>(0);
    const [open, setOpen] = useState<boolean>(false);
    const [rowData, setRowData] = useState<any>("");
    const classes = useStyle();

    const onScroll = (e: any) => {
        const bottom = Math.trunc(e.target.scrollHeight - e.target.scrollTop) === e.target.clientHeight;
        if (bottom) {
            setCount(count + 1);
        }
    }

    const timer = () => {
        setCount(count + 1)
    };

    function getSearchData(e: any) {
        if (e.target.value) {
            setDisplayData(true);
            const searchValue = data.filter((item: any) => {
                if (item.title.toLowerCase().includes(e.target.value.toLowerCase()) || item.author.toLowerCase().includes(e.target.value.toLowerCase())) {
                    return item
                }
            })
            setSearchData(searchValue);
        } else {
            setDisplayData(false);
        }
    }

    function apiCall() {
        axios({
            method: 'get',
            url: `https://hn.algolia.com/api/v1/search_by_date?tags=story&page=${count}`,
            responseType: 'stream'
        })
            .then((res) => {
                setData(data.concat(res.data.hits));
            })
            .catch((error) => {
                console.log(error);
            });
    };

    useEffect(() => {
        apiCall();
        const id = setInterval(timer, 10000);
        return () => clearInterval(id);
    }, [count]);

    const handleClose = () => {
        setOpen(false);
    }
    const handleOpen = (x: any) => {
        const findData: any = data.find((item: any) => item.objectID === x);
        setRowData(findData);
        setOpen(true);
    }

    return (
        <Typography component="div" className={classes.home}>
            <Navbar handleSearch={getSearchData} />
            <TableContainer component={Paper} className={classes.tableContainer} onScroll={onScroll}>
                <Table className={classes.table} aria-label="simple table">
                    <TableHead>
                        <TableRow data-testid="tableRow">
                            <TableCell align="center" className={classes.heading} data-testid="serialNo">Sr. No.</TableCell>
                            <TableCell align="center" className={classes.heading}>Title</TableCell>
                            <TableCell align="center" className={classes.heading}>URL</TableCell>
                            <TableCell align="center" className={classes.heading}>Created_at</TableCell>
                            <TableCell align="center" className={classes.heading}>Author</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {displayData ?
                            searchData.map((row: any, i) => {
                                return <TableRow key={i} className='tableRow' onClick={(x) => handleOpen(row.objectID)}>
                                    <TableCell align="center" className="url">{i + 1}</TableCell>
                                    <TableCell component="th" scope="row" align="center" className="title">
                                        {row.title}
                                    </TableCell>
                                    <TableCell align="center" className="url">{row.url}</TableCell>
                                    <TableCell align="center">{row.created_at}</TableCell>
                                    <TableCell align="center" className="author">{row.author}</TableCell>
                                </TableRow>
                            })
                            :
                            data.map((row: any, i) => {
                                return <TableRow key={i} className='tableRow' onClick={(x) => handleOpen(row.objectID)}>
                                    <TableCell align="center" className="url">{i + 1}</TableCell>
                                    <TableCell component="th" scope="row" align="center" className="title">
                                        {row.title}
                                    </TableCell>
                                    <TableCell align="center" className="url">{row.url}</TableCell>
                                    <TableCell align="center">{row.created_at}</TableCell>
                                    <TableCell align="center" className="author">{row.author}</TableCell>
                                </TableRow>
                            })
                        }
                    </TableBody>
                </Table>
            </TableContainer>
            <Modal
                open={open}
                onClose={handleClose}
                aria-labelledby="simple-modal-title"
                aria-describedby="simple-modal-description"
                className={classes.modalBox}
            >
                <div className={classes.paper}>
                    <p id="simple-modal-description">
                        {JSON.stringify(rowData)}
                    </p>
                </div>
            </Modal>
        </Typography>
    )
}

export default Home;
