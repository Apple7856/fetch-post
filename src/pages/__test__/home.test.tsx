import React from 'react';
import { act, cleanup, fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Home from '../Home';
import Navbar from '../../components/Navbar';
import axios from "axios";

afterEach(() => {
    cleanup();
});


describe("Home Component", () => {

    test("Heading Testing", () => {
        const home = render(<Home />);
        const serialNo = home.getByTestId('serialNo');
        expect(serialNo.innerHTML).toBe('Sr. No.');
    })

    test("Table Row Testing", () => {
        const { getByTestId } = render(<Home />);
        const tableRow = getByTestId('tableRow');
        expect(tableRow.children.length).toBe(5);
    })

    test("Match SnapShot", () => {
        const component = renderer.create(<Home />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('Search input value', () => {
        render(<Navbar handleSearch={() => jest.fn()} />);
        const searchValue: any = screen.getByTestId('searchBar').querySelector('input');
        fireEvent.change(searchValue, { target: { value: "SearchData" } })
        expect(searchValue.value).toBe("SearchData")
    })

})

describe('on useEffect Call Api for data', () => {
    it("should return users list", async () => {
        await act(async () => {
            const responce = await axios.get("https://hn.algolia.com/api/v1/search_by_date?tags=story&page=0");
            expect(responce.data.hits).toBeDefined();
        })
    });
})


