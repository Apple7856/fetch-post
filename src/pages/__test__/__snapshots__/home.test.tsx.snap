// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Match SnapShot 1`] = `
<div
  className="MuiTypography-root makeStyles-home-21 MuiTypography-body1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-27 MuiToolbar-gutters"
    >
      <h6
        className="MuiTypography-root makeStyles-logo-28 MuiTypography-h6"
      >
        Fetch-Post
      </h6>
      <div
        className="makeStyles-search-29"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <div
          className="MuiInputBase-root makeStyles-input-30"
          data-testid="searchBar"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search..."
            type="text"
            value="SearchData"
          />
        </div>
      </div>
    </div>
  </header>
  <div
    className="MuiPaper-root MuiTableContainer-root makeStyles-tableContainer-22 MuiPaper-elevation1 MuiPaper-rounded"
    onScroll={[Function]}
  >
    <table
      aria-label="simple table"
      className="MuiTable-root makeStyles-table-23"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          data-testid="tableRow"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-heading-24 MuiTableCell-alignCenter"
            data-testid="serialNo"
            scope="col"
          >
            Sr. No.
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-heading-24 MuiTableCell-alignCenter"
            data-testid="title"
            scope="col"
          >
            Title
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-heading-24 MuiTableCell-alignCenter"
            scope="col"
          >
            URL
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-heading-24 MuiTableCell-alignCenter"
            scope="col"
          >
            Created_at
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head makeStyles-heading-24 MuiTableCell-alignCenter"
            scope="col"
          >
            Author
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
</div>
`;
